ZEngine/src/Helper.cpp

```diff
@@ -0,0 +1,879 @@
+        ZENGINE_VALIDATE_ASSERT(
+            vkAllocateMemory(device.GetNativeDeviceHandle(), &memory_allocate_info, nullptr, &device_memory) == VK_SUCCESS, "Failed to allocate memory for vertex buffer")
+        ZENGINE_VALIDATE_ASSERT(vkBindBufferMemory(device.GetNativeDeviceHandle(), buffer, device_memory, 0) == VK_SUCCESS, "Failed to bind the memory to the vertex buffer")
+    }
+
+    void CopyBuffer(Hardwares::VulkanDevice& device, VkBuffer& source_buffer, VkBuffer& destination_buffer, VkDeviceSize byte_size)
+    {
+        auto            transfer_queue_view = device.GetAnyTransferQueue();
+        auto            command_pool        = device.GetTransferCommandPool(transfer_queue_view.FamilyIndex);
+        VkCommandBuffer command_buffer      = BeginOneTimeCommandBuffer(device, command_pool);
+
+        VkBufferCopy buffer_copy = {};
+        buffer_copy.srcOffset    = 0;
+        buffer_copy.dstOffset    = 0;
+        buffer_copy.size         = byte_size;
+
+        vkCmdCopyBuffer(command_buffer, source_buffer, destination_buffer, 1, &buffer_copy);
+
+        EndOneTimeCommandBuffer(command_buffer, device, transfer_queue_view.Queue, command_pool);
+    }
+
+    VkCommandBuffer BeginOneTimeCommandBuffer(Hardwares::VulkanDevice& device, VkCommandPool pool)
+    {
+        VkCommandBufferAllocateInfo command_buffer_allocate_info = {};
+        command_buffer_allocate_info.sType                       = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+        command_buffer_allocate_info.level                       = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+        command_buffer_allocate_info.commandBufferCount          = 1;
+        command_buffer_allocate_info.commandPool                 = pool;
+
+        VkCommandBuffer command_buffer;
+        ZENGINE_VALIDATE_ASSERT(vkAllocateCommandBuffers(device.GetNativeDeviceHandle(), &command_buffer_allocate_info, &command_buffer) == VK_SUCCESS, "")
+
+        VkCommandBufferBeginInfo command_buffer_begin_info = {};
+        command_buffer_begin_info.sType                    = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+        command_buffer_begin_info.flags                    = VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT;
+
+        ZENGINE_VALIDATE_ASSERT(vkBeginCommandBuffer(command_buffer, &command_buffer_begin_info) == VK_SUCCESS, "Failed to begin the Command Buffer")
+        return command_buffer;
+    }
+
+    void EndOneTimeCommandBuffer(VkCommandBuffer& command_buffer, Hardwares::VulkanDevice& device, VkQueue queue, VkCommandPool pool)
+    {
+        ZENGINE_VALIDATE_ASSERT(vkEndCommandBuffer(command_buffer) == VK_SUCCESS, "Failed to end the Command Buffer")
+
+        VkSubmitInfo submit_info       = {};
+        submit_info.sType              = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+        submit_info.commandBufferCount = 1;
+        submit_info.pCommandBuffers    = &command_buffer;
+
+        vkQueueSubmit(queue, 1, &submit_info, VK_NULL_HANDLE);
+        vkQueueWaitIdle(queue);
```

Comment:
I think we talked about ...WaitIdle already, so just for completeness: These will block the CPU until the queue is emptied and also will reduce GPU occupancy, since the queues might never fill up with command buffers. It is basically lockstepping the CPU and GPU, so the better way would be to submit the work and check via synchronization objects like fences, timeline semaphores, events, etc. if the transfer has completed and the object is usable.

Reply:
Thanks a lot !
I removed the use of vkQueueWaitIdle and switch to Fence and Semaphore